File: typescript/src/retrieval.ts
/**
 * Pure memory scoring and ranking, independent of storage or embeddings.
 *
 * The relevance x recency x salience decomposition follows Generative Agents
 * (Park et al. 2023), which sums those factors where this module multiplies
 * them, and decays recency exponentially where this module is linear to a
 * floor. Constant provenance is recorded in docs/foundations.md sections 6-7.
 */
import type { MemoryCandidate, RankedMemory } from './contracts.js';

function clamp(value: number, low: number, high: number): number {
  return Math.max(low, Math.min(high, value));
}

/**
 * The numeric parameters of the retrieval scorer.
 *
 * Defaults reproduce the pinned cross-runtime contract. These are magnitudes
 * only: the *shape* of the curves -- linear recency, multiplicative
 * composition, log-compressed rehearsal -- is fixed here, and
 * docs/foundations.md section 6 records why each shape was chosen and what the
 * closest published comparable does instead.
 */
export interface RetrievalWeights {
  readonly recencyHorizonDays: number;
  readonly recencyFloor: number;
  readonly recencyFallback: number;
  readonly significanceWeight: number;
  readonly rehearsalWeight: number;
  readonly rehearsalCap: number;
  readonly episodeBonus: number;
  readonly congruenceThreshold: number;
  readonly congruenceNegativeMood: number;
  readonly congruencePositiveMood: number;
}

export const DEFAULT_RETRIEVAL_WEIGHTS: RetrievalWeights = Object.freeze({
  recencyHorizonDays: 60,
  recencyFloor: 0.4,
  recencyFallback: 0.7,
  significanceWeight: 0.03,
  rehearsalWeight: 0.006,
  rehearsalCap: 0.025,
  episodeBonus: 0.05,
  congruenceThreshold: 0.2,
  congruenceNegativeMood: 1.06,
  congruencePositiveMood: 1.03,
});

export function recencyWeight(
  daysAgo: number,
  weights: RetrievalWeights = DEFAULT_RETRIEVAL_WEIGHTS,
): number {
  if (!Number.isFinite(daysAgo)) throw new TypeError('daysAgo must be finite');
  return clamp(1 - daysAgo / weights.recencyHorizonDays, weights.recencyFloor, 1);
}

export function recencyWeightFromTimestamp(
  timestamp: string,
  now = new Date(),
  weights: RetrievalWeights = DEFAULT_RETRIEVAL_WEIGHTS,
): number {
  const time = Date.parse(timestamp);
  if (!Number.isFinite(time)) return weights.recencyFallback;
  return recencyWeight((now.getTime() - time) / 86_400_000, weights);
}

/**
 * Mood-congruent recall is Bower (1981); the threshold, the magnitudes, and the
 * negative/positive asymmetry are production-tuned and unsupported by any
 * citation in docs/foundations.md section 7.
 */
export function moodCongruenceFactor(
  memoryValence: number,
  moodValence: number,
  congruenceOn: boolean,
  weights: RetrievalWeights = DEFAULT_RETRIEVAL_WEIGHTS,
): number {
  if (!congruenceOn || memoryValence === 0) return 1;
  if ((memoryValence > 0) === (moodValence > 0)) {
    return moodValence < 0 ? weights.congruenceNegativeMood : weights.congruencePositiveMood;
  }
  return 1;
}

export function similarityFromDistance(distance: number): number {
  if (!Number.isFinite(distance)) throw new TypeError('distance must be finite');
  return clamp(1 - distance / 2, 0, 1);
}

export interface CandidateScoreInput {
  similarity: number;
  recency: number;
  episode: boolean;
  significance: number;
  recallCount: number;
  weights?: RetrievalWeights;
}

export function candidateScore(input: CandidateScoreInput): number {
  if (!Number.isFinite(input.similarity)) throw new TypeError('similarity must be finite');
  if (!Number.isFinite(input.recency)) throw new TypeError('recency must be finite');
  if (!Number.isFinite(input.significance)) throw new TypeError('significance must be finite');
  if (!Number.isSafeInteger(input.recallCount) || input.recallCount < 0) {
    throw new RangeError('recallCount must be a non-negative safe integer');
  }
  const weights = input.weights ?? DEFAULT_RETRIEVAL_WEIGHTS;
  const significance = clamp(input.significance, 0, 1);
  const recallCount = input.recallCount;
  const salience = 1 + significance * weights.significanceWeight
    + Math.min(weights.rehearsalCap, Math.log1p(recallCount) * weights.rehearsalWeight);
  return clamp(input.similarity, 0, 1) * clamp(input.recency, weights.recencyFloor, 1) * salience
    + (input.episode ? weights.episodeBonus : 0);
}

export interface RankCandidatesOptions {
  limit?: number;
  now?: Date;
  moodValence?: number;
  moodCongruence?: boolean;
  weights?: RetrievalWeights;
}

const ZONED_TIMESTAMP = /T.*(?:Z|[+-]\d{2}:\d{2})$/iu;

function validateCandidate(candidate: MemoryCandidate): void {
  if (typeof candidate.id !== 'string' || candidate.id.length === 0) {
    throw new RangeError('memory candidate id must not be empty');
  }
  if (typeof candidate.text !== 'string') throw new TypeError('memory candidate text must be a string');
  if (!Number.isFinite(candidate.distance) || candidate.distance < 0 || candidate.distance > 2) {
    throw new RangeError('memory candidate distance must be finite and between 0 and 2');
  }
  if (candidate.daysAgo !== undefined
    && (!Number.isFinite(candidate.daysAgo) || candidate.daysAgo < 0)) {
    throw new RangeError('memory candidate daysAgo must be finite and non-negative');
  }
  if (candidate.timestamp !== undefined && candidate.timestamp !== null) {
    if (typeof candidate.timestamp !== 'string') {
      throw new TypeError('memory candidate timestamp must be a string or null');
    }
    if (!ZONED_TIMESTAMP.test(candidate.timestamp) || !Number.isFinite(Date.parse(candidate.timestamp))) {
      throw new RangeError('memory candidate timestamp must be a valid zoned ISO timestamp');
    }
  }
  if (candidate.episode !== undefined && typeof candidate.episode !== 'boolean') {
    throw new TypeError('memory candidate episode must be a boolean');
  }
  if (candidate.significance !== undefined
    && (!Number.isFinite(candidate.significance)
      || candidate.significance < 0
      || candidate.significance > 1)) {
    throw new RangeError('memory candidate significance must be between 0 and 1');
  }
  if (candidate.recallCount !== undefined
    && (!Number.isSafeInteger(candidate.recallCount) || candidate.recallCount < 0)) {
    throw new RangeError('memory candidate recallCount must be a non-negative safe integer');
  }
  if (candidate.emotionalValence !== undefined
    && (!Number.isFinite(candidate.emotionalValence)
      || candidate.emotionalValence < -1
      || candidate.emotionalValence > 1)) {
    throw new RangeError('memory candidate emotionalValence must be between -1 and 1');
  }
}

export function scoreCandidate(
  candidate: MemoryCandidate,
  options: Omit<RankCandidatesOptions, 'limit'> = {},
): number {
  validateCandidate(candidate);
  const now = options.now ?? new Date();
  const weights = options.weights ?? DEFAULT_RETRIEVAL_WEIGHTS;
  const recency = candidate.daysAgo === undefined
    ? recencyWeightFromTimestamp(candidate.timestamp ?? '', now, weights)
    : recencyWeight(candidate.daysAgo, weights);
  const score = candidateScore({
    similarity: similarityFromDistance(candidate.distance ?? 0),
    recency,
    episode: candidate.episode ?? false,
    significance: candidate.significance ?? 0.5,
    recallCount: candidate.recallCount ?? 0,
    weights,
  });
  const moodValence = options.moodValence ?? 0;
  const congruenceOn =
    (options.moodCongruence ?? true) && Math.abs(moodValence) >= weights.congruenceThreshold;
  return score * moodCongruenceFactor(
    candidate.emotionalValence ?? 0,
    moodValence,
    congruenceOn,
    weights,
  );
}

export function rankCandidates(
  candidates: Iterable<MemoryCandidate>,
  options: RankCandidatesOptions = {},
): RankedMemory[] {
  const limit = options.limit ?? 4;
  if (!Number.isInteger(limit) || limit < 0) {
    throw new RangeError('limit must be a non-negative integer');
  }
  const best = new Map<string, RankedMemory>();
  for (const candidate of candidates) {
    const ranked = { ...candidate, score: scoreCandidate(candidate, options) };
    const previous = best.get(candidate.id);
    if (!previous || ranked.score > previous.score) best.set(candidate.id, ranked);
  }
  return [...best.values()]
    .sort((left, right) => right.score - left.score || compareCodePoints(left.id, right.id))
    .slice(0, limit);
}

function compareCodePoints(left: string, right: string): number {
  const leftIterator = left[Symbol.iterator]();
  const rightIterator = right[Symbol.iterator]();
  while (true) {
    const leftNext = leftIterator.next();
    const rightNext = rightIterator.next();
    if (leftNext.done || rightNext.done) {
      if (leftNext.done === rightNext.done) return 0;
      return leftNext.done ? -1 : 1;
    }
    const leftPoint = leftNext.value.codePointAt(0) as number;
    const rightPoint = rightNext.value.codePointAt(0) as number;
    if (leftPoint !== rightPoint) return leftPoint < rightPoint ? -1 : 1;
  }
}

File: python/src/affect_kernel/retrieval.py
"""Pure memory scoring and ranking, independent of storage or embeddings.

The relevance x recency x salience decomposition follows Generative Agents
(Park et al. 2023), which sums those factors where this module multiplies them.
Constant provenance is recorded in ``docs/foundations.md`` sections 6-7.
"""

from __future__ import annotations

import math
from collections.abc import Iterable
from dataclasses import dataclass
from datetime import UTC, datetime

from .models import MemoryCandidate, RankedMemory


def _bounded_number(name: str, value: float, low: float, high: float) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise TypeError(f"{name} must be a number")
    number = float(value)
    if not math.isfinite(number) or not low <= number <= high:
        raise ValueError(f"{name} must be finite and within [{low}, {high}]")
    return number


@dataclass(frozen=True, slots=True)
class RetrievalWeights:
    """The numeric parameters of the retrieval scorer.

    Defaults reproduce the pinned cross-runtime contract. These are magnitudes
    only: the *shape* of the curves -- linear recency, multiplicative
    composition, log-compressed rehearsal -- is fixed here, and
    ``docs/foundations.md`` section 6 records why each shape was chosen and what
    the closest published comparable does instead.
    """

    recency_horizon_days: float = 60.0
    recency_floor: float = 0.40
    recency_fallback: float = 0.70
    significance_weight: float = 0.03
    rehearsal_weight: float = 0.006
    rehearsal_cap: float = 0.025
    episode_bonus: float = 0.05
    congruence_threshold: float = 0.20
    congruence_negative_mood: float = 1.06
    congruence_positive_mood: float = 1.03

    def __post_init__(self) -> None:
        horizon = _bounded_number(
            "recency_horizon_days", self.recency_horizon_days, 0.0, 3_650_000.0
        )
        if horizon <= 0.0:
            raise ValueError("recency_horizon_days must be positive")
        for name in (
            "recency_floor",
            "recency_fallback",
            "significance_weight",
            "rehearsal_weight",
            "rehearsal_cap",
            "episode_bonus",
            "congruence_threshold",
        ):
            _bounded_number(name, getattr(self, name), 0.0, 1.0)
        for name in ("congruence_negative_mood", "congruence_positive_mood"):
            _bounded_number(name, getattr(self, name), 0.0, 10.0)


DEFAULT_RETRIEVAL_WEIGHTS = RetrievalWeights()


def _require_aware(value: datetime, name: str) -> datetime:
    if value.tzinfo is None or value.utcoffset() is None:
        raise ValueError(f"{name} must be timezone-aware")
    return value


def _validate_distance(distance: float) -> float:
    if isinstance(distance, bool) or not isinstance(distance, (int, float)):
        raise TypeError("distance must be a number")
    if not math.isfinite(distance) or not 0.0 <= distance <= 2.0:
        raise ValueError("distance must be finite and within [0, 2]")
    return distance


def recency_weight(
    timestamp: str,
    *,
    now: datetime | None = None,
    weights: RetrievalWeights | None = None,
) -> float:
    """Return a linear freshness weight with a 0.4 floor and 0.7 parse fallback.

    Human forgetting is better described by a power law (Wixted & Ebbesen 1991)
    and the closest published comparable decays exponentially, so this curve is
    the one most obviously open to challenge. It survived that challenge: at a
    matched 30-day half-life it out-ranked both alternatives in ``bench/``. See
    ``docs/foundations.md`` section 6.
    """
    reference = _require_aware(now or datetime.now(UTC), "now")
    tuning = weights or DEFAULT_RETRIEVAL_WEIGHTS
    try:
        parsed = datetime.fromisoformat(timestamp)
        _require_aware(parsed, "timestamp")
        days_ago = max(0.0, (reference - parsed).total_seconds() / 86_400)
        return max(
            tuning.recency_floor,
            min(1.0, 1.0 - days_ago / tuning.recency_horizon_days),
        )
    except (TypeError, ValueError, OverflowError):
        return tuning.recency_fallback


def mood_congruence_factor(
    mem_valence: float,
    mood_valence: float,
    congruence_on: bool,
    *,
    weights: RetrievalWeights | None = None,
) -> float:
    """Return the small asymmetric multiplier for same-sign memory and mood valence.

    Mood-congruent recall is Bower (1981); the threshold, the magnitudes, and
    the negative/positive asymmetry are production-tuned and unsupported by any
    citation in ``docs/foundations.md`` section 7.
    """
    if not congruence_on or mem_valence == 0.0:
        return 1.0
    tuning = weights or DEFAULT_RETRIEVAL_WEIGHTS
    if (mem_valence > 0.0) == (mood_valence > 0.0):
        return (
            tuning.congruence_negative_mood
            if mood_valence < 0.0
            else tuning.congruence_positive_mood
        )
    return 1.0


def similarity_from_distance(distance: float) -> float:
    """Convert cosine distance from the production convention into similarity."""
    return 1.0 - _validate_distance(distance) / 2.0


def candidate_score(
    distance: float,
    days_ago: float,
    *,
    episode: bool,
    significance: float,
    recall_count: int,
    weights: RetrievalWeights | None = None,
) -> float:
    """Score a worked candidate from distance, age, salience, and rehearsal."""
    _validate_distance(distance)
    if not math.isfinite(days_ago):
        raise ValueError("days_ago must be finite")
    if not math.isfinite(significance):
        raise ValueError("significance must be finite")
    if isinstance(recall_count, bool) or not isinstance(recall_count, int):
        raise TypeError("recall_count must be an integer")
    if recall_count < 0:
        raise ValueError("recall_count must be non-negative")
    tuning = weights or DEFAULT_RETRIEVAL_WEIGHTS
    similarity = similarity_from_distance(distance)
    recency = max(
        tuning.recency_floor,
        min(1.0, 1.0 - max(0.0, days_ago) / tuning.recency_horizon_days),
    )
    bounded_significance = max(0.0, min(1.0, significance))
    salience = (
        1.0
        + bounded_significance * tuning.significance_weight
        + min(tuning.rehearsal_cap, math.log1p(recall_count) * tuning.rehearsal_weight)
    )
    return similarity * recency * salience + (tuning.episode_bonus if episode else 0.0)


def score_candidate(
    candidate: MemoryCandidate,
    *,
    now: datetime | None = None,
    mood_valence: float = 0.0,
    mood_congruence: bool = True,
    weights: RetrievalWeights | None = None,
) -> float:
    """Score a candidate carrying an ISO timestamp and optional emotional valence."""
    reference = _require_aware(now or datetime.now(UTC), "now")
    tuning = weights or DEFAULT_RETRIEVAL_WEIGHTS
    similarity = similarity_from_distance(candidate.distance)
    recency = recency_weight(candidate.timestamp or "", now=reference, weights=tuning)
    significance = max(0.0, min(1.0, candidate.significance))
    salience = (
        1.0
        + significance * tuning.significance_weight
        + min(
            tuning.rehearsal_cap,
            math.log1p(max(0, candidate.recall_count)) * tuning.rehearsal_weight,
        )
    )
    score = similarity * recency * salience + (tuning.episode_bonus if candidate.episode else 0.0)
    congruence_on = mood_congruence and abs(mood_valence) >= tuning.congruence_threshold
    return score * mood_congruence_factor(
        candidate.emotional_valence,
        mood_valence,
        congruence_on,
        weights=tuning,
    )


def rank_candidates(
    candidates: Iterable[MemoryCandidate],
    *,
    limit: int = 4,
    now: datetime | None = None,
    mood_valence: float = 0.0,
    mood_congruence: bool = True,
    weights: RetrievalWeights | None = None,
) -> tuple[RankedMemory, ...]:
    """Deduplicate by id, retain the best score, and return a stable descending rank."""
    if limit < 0:
        raise ValueError("limit must be non-negative")
    reference = _require_aware(now or datetime.now(UTC), "now")
    best: dict[str, RankedMemory] = {}
    for candidate in candidates:
        ranked = RankedMemory(
            candidate=candidate,
            score=score_candidate(
                candidate,
                now=reference,
                mood_valence=mood_valence,
                mood_congruence=mood_congruence,
                weights=weights,
            ),
        )
        previous = best.get(candidate.id)
        if previous is None or ranked.score > previous.score:
            best[candidate.id] = ranked
    ordered = sorted(best.values(), key=lambda item: (-item.score, item.candidate.id))
    return tuple(ordered[:limit])


__all__ = [
    "DEFAULT_RETRIEVAL_WEIGHTS",
    "RetrievalWeights",
    "candidate_score",
    "mood_congruence_factor",
    "rank_candidates",
    "recency_weight",
    "score_candidate",
    "similarity_from_distance",
]
